src/Cryptol/REPL/Command.hs

```diff
@@ -917,50 +920,66 @@ offlineProveSat proverName qtype expr schema mfile = do
         , pcSchema       = schema
         , pcIgnoreSafety = ignoreSafety
         }
-
   put <- getPutStr
-  let putLn x = put (x ++ "\n")
-  let displayMsg =
-        do let filename = fromMaybe "standard output" mfile
-           let satWord = case qtype of
-                           SatQuery _  -> "satisfiability"
-                           ProveQuery  -> "validity"
-                           SafetyQuery -> "safety"
-           putLn $
-               "Writing to SMT-Lib file " ++ filename ++ "..."
-           putLn $
-             "To determine the " ++ satWord ++
-             " of the expression, use an external SMT solver."
 
   getProverConfig >>= \case
     Left sbvCfg ->
       do result <- liftModuleCmd $ SBV.satProveOffline sbvCfg cmd
          case result of
            Left msg -> rPutStrLn msg
            Right smtlib -> do
-             io $ displayMsg
+             io $ displayWriteSMTLibMsg qtype mfile put
              case mfile of
                Just path -> io $ writeFile path smtlib
                Nothing -> rPutStr smtlib
 
     Right _w4Cfg ->
       do ~(EnvBool hashConsing) <- getUser "hashConsing"
          ~(EnvBool warnUninterp) <- getUser "warnUninterp"
-         result <- liftModuleCmd $ W4.satProveOffline hashConsing warnUninterp cmd $ \f ->
-                     do displayMsg
-                        case mfile of
-                          Just path ->
-                            X.bracket (openFile path WriteMode) hClose f
-                          Nothing ->
-                            withRWTempFile "smtOutput.tmp" $ \h ->
-                              do f h
-                                 hSeek h AbsoluteSeek 0
-                                 hGetContents h >>= put
+         result <- liftModuleCmd $ W4.satProveOffline hashConsing warnUninterp cmd
+                                 $ w4WriteSMTLibOutput qtype mfile put
 
          case result of
            Just msg -> rPutStrLn msg
            Nothing -> return ()
 
+-- | Write SMT-Lib output using a What4-based backend.
+w4WriteSMTLibOutput ::
+  QueryType ->
+  Maybe FilePath ->
+  (String -> IO ()) {- ^ Continuation for displaying user messages -} ->
+  (Handle -> IO ()) {- ^ Continuation for writing file output -} ->
```

Comment:
I think a more traditional name for these are "callbacks" rather than "continuations"?